### DISCUSSION TITLE
A way to visualize system transformations

### DISCUSSION THREAD

**Discussion by susliko** [likec4/likec4#1192] [Ideas]
Use case: "An engineer wants to create a design doc on system transformation to support the implementation of a new feature. They can add new system elements and relationships into a new file  `feature.c4` and assemble views that may rely on both existing and new elements and relationships. Other views of the workspace should not be affected."

The non-elegant solution I have found is as follows. New ("to be") elements and relations are marked with tags  `#future` and `#featureX`. Every "as is" view excludes `#future` tag. Feature-specific "to be" views use includes with filters on `#featureX` tag.

The problem with this approach is that every "as is" view must contain 
```
exclude element where tag is #future
exclude * -> * where tag is #future
```

Another approach with similar shortcomings is tagging each and every component and relationship and constructing views by these tags.

Potential solution i can think of is to introduce visibility levels like programming languages do. A simple (from the perspective of keeping language semantics simple to grasp) way to achieve this might be to allow file-local model extensions.

For example

```
# file my-system.c4
model {
  my-system = system "My System { ... }
}
views {
   landscape of my-system {
      # new-container is not accessible here
   }
}  
```

```
# file features/featureX.c4

model {
  local extend my-model {
     new-container = container "New container" { ... }
  }
}

view {
  include my-system.new-container
}
```

**Comment by davydkov** [comment]
Wow, what a coincidence! System evolution, transformation, versioning, and pinned views are all topics on my mind. I’m working on combining these ideas into an RFC for publication.

The idea of visibility levels wasn’t on my radar yet, but I think it’s definitely worth to consider — thanks for the suggestion!

In the meantime, take a look at the upcoming #1173 
Does it address your needs?

In a nutshell:
```zig
global {
  predicateGroup new_cloud_service {
    include cloud.*
      where kind is microservice
    exclude *
      where tag is #deprecated
  }
}
views {
  view of newServices {
    include *
    global predicate new_cloud_service
  }
  view of newBackendServices {
    include *
    global predicate new_cloud_service
  }
}
```

**Reply by susliko** [reply]
> I’m working on combining these ideas into an RFC for publication.

It's great to hear that!

As for global predicates, I think they could help reduce some boilerplate but might not fully address the core issue, as their use for system change visualization would still require careful tagging.

Perhaps what we need here is some feature with a 'stability' property with respect to existing schemas. The one that would allow us to add ad-hoc views without affecting existing ones while avoiding a full workspace revision to apply tags or filters.

**Reply by susliko** [reply]
I don't know if it's of any use, but in my head the concept looks like this 😅 
I'd like to be able to extend existing model and assemble views for "RFC 1" and "RFC 2" so that these modifications are not even accessible from "Architecture As Is".
![image](https://github.com/user-attachments/assets/c93aa2b0-6d69-489d-9aa4-b244864d4047)

These rfc-related modifications are temporary. They will be either discarded (if RFC gets rejected) or incorporated into "Architecture As Is" upon implementation.

**Comment by Cyclonit** [comment]
Maybe a system like git diffs could address this feature request. Consider the following example based on the [bigbank example](https://playground.likec4.dev/w/bigbank/index/):

```
diff {
  remove {
    // removes the element "bigbank.supportStaff" and all edges connecting to it
    bigbank.supportStaff

    // remove the edge from internetBankingSystem to email
    bigbank.internetBankingSystem -> bigbank.email
  }

  add {
    // create a new element in the bigbank enterprise
    bigbank.moneyDeliveryStaff = staff "Money Delivery Staff" {
      description "Staff delivering money to customers."
    }
  }
}
```

pros:

- **modularization**
Extracting change sets into separate files avoids the mess of having actual and planned infrastructure crammed into one model.

- **validation**
Because the change set is described as a distinct entity, it is possible to check if it fits the current model. If something in the architecture changes while the RFC is being discussed, it would be trivial to check if the RFC requires updating.

- **visualization**
The viewer could extended to visualize what happens when applying a set of changes.

cons:

- **integration**  
When an RFC/change is accepted, its change set should be applied to the actual model. It is possible to automate this using the CLI. However, because there are several ways of describing the same architecture using the DSL, this might not result in the exact code structure the user wants it to.

**Reply by susliko** [reply]
The idea of using cli for accepting/rejecting such patches is pretty cool

**Reply by windead** [reply]
It would be amazing if likec4 could accept git diff as an input and visualize the differences directly in the diagram view.

Here’s why this feature would be valuable:

Efficient Change Review: When managing design documents with Git, it’s often necessary to review changes between different versions. By integrating git diff support, likec4 could provide a clear and intuitive visual comparison of design modifications.
Streamlined Workflow: This feature would seamlessly integrate with the Git workflow, enabling users to pull two different versions of a design document and quickly understand the updates without leaving the likec4 environment.
Enhanced Collaboration: Visualizing diffs directly in diagrams could improve team collaboration, as team members could better understand the evolution of designs and provide targeted feedback.
Suggested Implementation:

Input Format: Accept git diff output as an input, either by specifying file paths or by directly integrating with Git repositories.
Visualization: Highlight added, removed, or modified components in the diagrams, using color codes (e.g., green for additions, red for deletions, yellow for modifications).
Version History Support: Provide an option to navigate through commit histories to visualize design changes over time.
I believe this feature aligns with likec4’s goal of making software architecture visualization more accessible and practical for developers.

**Reply by edeak-jsq** [reply]
I like this approach because to me it grabs the diff on the right level: at the model.
- we have a model in state S0
- we can enrich the model with a diff (add, remove)
- applying diff on S0 will take the model to S1 (the next state)
- S0 and S1 visualization is already supported
- diff visualization might require some additions on the views feature

I'd take this path.

**Comment by eriklenaerts** [comment]
hey all, I found a way to work with the existing language constructs to showcase architecture transitions. 
Here's the playground: https://playground.likec4.dev/share/mcRojIPxO9/

The idea is as follow: 
- we make use of #deprecated and #future tags
- for every view we have one As-Is and one To-Be, the latter extends the former
- by means of global predicates we hide or show elements and and relationships and highlight these changes (using styling)

```
global {
    predicateGroup as-is {
        exclude * -> * 
            where tag = #future
        exclude * 
            where tag = #future
        include * -> * 
            where tag = #depcrecated with {
                notation 'deprecated'
                color amber
            }
        include * 
            where tag = #depcrecated with {
                notation 'deprecated'
                color amber
            }
    }

    predicateGroup to-be {
        exclude * -> * 
            where tag = #depcrecated
        exclude * 
            where tag = #depcrecated
        include * -> * 
            where tag = #future with {
                notation 'future'
                color green
            }
        include * 
            where tag = #future with {
                notation 'future'
                color green
            }
    }
}
```

as we're still in early phase, I can't really say that this approach will keep working for us, though I figured to share this already ;)

**Reply by tib-rijo** [reply]
I'm interested in seeing more like this. The playground has meanwhile expired.

**Comment by jasondamour** [comment]
Maybe the solution is a concept which aggregates views, i.e. a "story" which is like sequenced views:
```
view 'monolith' {
  include * where tag is #monolith and tag is not #microservice
}

view 'microservices {
  include * where tag is #microservice and tag is not #monolith
}

story "migration to microservices" {
  monolith -[fade]-> microservice 
}
```

Then likec4 renders these in some dynamic fashion, maybe similar to a slide deck. The relationship from one view to another can be an enum of transitions like `cut` (hard cut from one view to another), or `fade` which briefly shows the models from both views to represent "gradual" migrations.

**Reply by jasondamour** [reply]
@davydkov, I liked this concept so much that I started proving it out!
- [RFC here](https://github.com/jasondamour/likec4/blob/story-view-implementation/docs/rfcs/0001-story-view.md)
- [Initial Implementation here](https://github.com/likec4/likec4/pull/3172)
- [Demo video (sorry about the compression)](https://github.com/user-attachments/assets/f98e0fb7-bb6e-4758-82df-43f5dd3f30c4)

TL;DR
- A new conceptual layer between projects and views called **Story**
  - `/project/abc/story/cloud-migration/view/status-quo`
- Stories are a sequence of Scenes
- A Scene is just a View, plus metadata about how it relates to other Views
  - i.e. what models should be "anchored" in place, what models should "become" other models, etc
  - In the future, stories can support all the flow keywords that dynamic views support (i.e. alt, loop, parallel, etc)

**Comment by StefanStuehrmann** [comment]
Would be amazing to work on this. I started in some projects to create a separate projects for vision or target state but it's also kind of messy as you are now maintaining 2 projects and you need to move things around a lot.
